### PR TITLE
A more robust implementation of out parameters in the interface

### PR DESCRIPTION
**Pull request by eggrobin**
Allow out parameters of types which don't have 0 as a value, in preparation for #901.
Require that out parameters be [`TriviallyCopyable`](http://en.cppreference.com/w/cpp/concept/TriviallyCopyable), since otherwise they cannot be truly out (note: could we get away with just trivial assignments and destructor?).
Enforce that they be truly out by stomping the memory before passing them.

The use of `make_unique` means we also require a (not necessarily trivial) default constructor, but this could easily be circumvented using stuff from from [`<new>`](http://en.cppreference.com/w/cpp/header/new), or `malloc` and a `unique_ptr` with `free` as its custom deleter, if needed.

See the changes to the generated code at 5f631b4.

### CODE REVIEW

File: tools/journal_proto_processor.cpp [hunk 1]
```diff
@@ -481,10 +488,17 @@ void JournalProtoProcessor::ProcessRequiredField(
     }
     field_cxx_type_[descriptor] += "*";
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: We should have two wrapper lambdas `field_cxx_mode_fn_` and `field_cs_mode_fn_`.

File: tools/journal_proto_processor.cpp [hunk 2]
```diff
@@ -481,10 +488,17 @@ void JournalProtoProcessor::ProcessRequiredField(
...
     field_cxx_arguments_fn_[descriptor] =
         [](std::string const& identifier) -> std::vector<std::string> {
           return {"&" + identifier};
         };
+    } else {
+      field_cxx_arguments_fn_[descriptor] =
+        [](std::string const& identifier) -> std::vector<std::string> {
+        return { identifier + ".get()" };
+      };
+    }
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: OBE.

File: tools/journal_proto_processor.cpp [hunk 3]
```diff
@@ -606,24 +620,55 @@ void JournalProtoProcessor::ProcessInOut(
...
-          ";\n";
+      if (Contains(out_, field_descriptor)) {
+        // For an out parameter, we need to allocate some uninitializeed
+        // storage.  The |make_unique| below requires that
+        // |field_cxx_type_[field_descriptor]| be default-constructible.
+        // For more generality, we could initialize the |unique_ptr| to
+        // |reinterpret_cast<T*>(std::operator new(sizeof(T)))|.
+        cxx_run_body_prolog_[descriptor] +=
+            "  auto const " + run_local_variable + " = std::make_unique<" +
+            field_cxx_direct_type_[field_descriptor] + ">();\n";
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: No need for a `unique_ptr<>`, this should generate something like:
```
  T t;
```
where `T` is `field_cxx_type_`.

File: tools/journal_proto_processor.cpp [hunk 4]
```diff
@@ -606,24 +620,55 @@ void JournalProtoProcessor::ProcessInOut(
...
+        cxx_run_body_prolog_[descriptor] +=
+            "  std::memset(" + run_local_variable + ".get(), 0xAB, sizeof (" +
+            field_cxx_direct_type_[field_descriptor] + "));\n";
+        cxx_run_body_prolog_[descriptor] +=
+            "  static_assert(\n    std::is_trivially_copyable<" +
+            field_cxx_direct_type_[field_descriptor] +
+            ">::value,\n    \"out parameter |" +
+            run_local_variable + "| of |" +
+            field_descriptor->containing_type()->containing_type()->name() +
+            "| must have a trivially copyable type\");\n";
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: Let's remove the stomping and the `static_assert` and have tighter `static_assert`s in the interface.

File: tools/journal_proto_processor.cpp [hunk 5]
```diff
@@ -606,24 +620,55 @@ void JournalProtoProcessor::ProcessInOut(
...
+      } else {
+        std::string const cxx_run_field_deserializer_getter =
+            cxx_run_field_getter;
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: This variable is no longer needed, it was just here to isolate a `?:` expression.

File: tools/journal_proto_processor.cpp [hunk 6]
```diff
@@ -606,24 +620,55 @@ void JournalProtoProcessor::ProcessInOut(
...
       cxx_run_body_epilog_[descriptor] +=
           field_cxx_deleter_fn_[field_descriptor](cxx_run_field_getter);
     }
     if (Contains(field_cxx_inserter_fn_, field_descriptor)) {
+      bool const is_out = Contains(out_, field_descriptor) &&
+                          !Contains(in_out_, field_descriptor);
       cxx_run_body_epilog_[descriptor] +=
           field_cxx_inserter_fn_[field_descriptor](
               ToLower(name) + "." + field_descriptor_name + "()",
+              (is_out ? "*" : "") +
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: OBE.

File: tools/journal_proto_processor.hpp [hunk 7]
```diff
@@ -182,6 +182,9 @@ class JournalProtoProcessor {
   // declaration, in a typedef, etc.
   std::map<FieldDescriptor const*, std::string> field_cs_type_;
   std::map<FieldDescriptor const*, std::string> field_cxx_type_;
+  // For out or in out parameters, the type of the parameter without the
+  // indirection due to the mode.
+  std::map<FieldDescriptor const*, std::string> field_cxx_direct_type_;
```
Comment (pleroy) [inline tools/journal_proto_processor.hpp]: OBE.

**Comment by eggrobin** [comment]
Changes to generated code now at 2df1b36.

**Comment by eggrobin** [comment]
Changes to the generated code now at e6c02c5.

File: tools/journal_proto_processor.cpp [hunk 8]
```diff
@@ -601,20 +624,19 @@ void JournalProtoProcessor::ProcessInOut(
                 std::back_inserter(cxx_run_arguments_[descriptor]));
     }
     if (must_generate_code) {
```
Comment (pleroy) [inline tools/journal_proto_processor.cpp]: Merge this `if` with the previous one.

File: tools/journal_proto_processor.hpp [hunk 9]
```diff
@@ -149,6 +149,18 @@ class JournalProtoProcessor {
                                      std::string const& expr2)>>
       field_cxx_inserter_fn_;
 
+  // For all fields, a lambda that takes a C#/C++ parameter type as stored in
+  // |field_cs_type_|, repsectively |field_cxx_type_|, and adds a mode to it,
```
Comment (pleroy) [inline tools/journal_proto_processor.hpp]: respectively

File: tools/journal_proto_processor.hpp [hunk 10]
```diff
@@ -149,6 +149,18 @@ class JournalProtoProcessor {
                                      std::string const& expr2)>>
       field_cxx_inserter_fn_;
 
+  // For all fields, a lambda that takes a C#/C++ parameter type as stored in
```
Comment (pleroy) [inline tools/journal_proto_processor.hpp]: parameter or member

File: tools/journal_proto_processor.hpp [hunk 11]
```diff
@@ -149,6 +149,18 @@ class JournalProtoProcessor {
...
+  // |field_cs_type_|, repsectively |field_cxx_type_|, and adds a mode to it,
+  // yielding |T| for an in parameter in either language, |ref T| for an in out
```
Comment (pleroy) [inline tools/journal_proto_processor.hpp]: I'd drop everything from "yielding" to the end, no point in leaking implementation details.

**Pull request merged**